sdk/core/core/inc/az_json.h

```diff
@@ -182,81 +196,233 @@ AZ_NODISCARD az_result az_json_token_get_string(az_json_token const* token, az_s
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
  */
 AZ_NODISCARD az_result
-az_json_builder_append_object(az_json_builder* json_builder, az_span name, az_json_token token);
+az_json_builder_append_property_name(az_json_builder* json_builder, az_span name);
 
-/*
- * @brief az_json_builder_append_array_item appends an array item to the JSON buffer.
+/**
+ * @brief Appends a boolean value (as a JSON literal `true` or `false`).
  *
- * @param json_builder A pointer to an az_json_builder instance containing the buffer to append the
- * array item to.
- * @param token A pointer to the az_json_token representing the array item.
- * @return AZ_OK if the token was appended successfully.<br>
- * AZ_ERROR_BUFFER_OVERFLOW if the buffer is too small.
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the boolean to.
+ * @param[in] value The value to be written as a JSON literal true or false.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if the boolean was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
+ */
+AZ_NODISCARD az_result az_json_builder_append_bool(az_json_builder* json_builder, bool value);
+
+// TODO: Consider removing for now until we have complete double formatting and parsing.
+/**
+ * @brief Appends a double number value.
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the number to.
+ * @param[in] value The value to be written as a JSON number.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if the number was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
+ */
+AZ_NODISCARD az_result az_json_builder_append_number(az_json_builder* json_builder, double value);
+
+/**
+ * @brief Appends an int32_t number value.
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the number to.
+ * @param[in] value The value to be written as a JSON number.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if the number was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
  */
 AZ_NODISCARD az_result
-az_json_builder_append_array_item(az_json_builder* json_builder, az_json_token token);
+az_json_builder_append_int32_number(az_json_builder* json_builder, int32_t value);
+
+/**
+ * @brief Appends the JSON literal `null`.
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the `null` literal to.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if `null` was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
```

Comment:
the problem of listing all the posible errors is that it creates a compromise to not adding more than those in the future.
I learned this from @JeffreyRichter 

So, if we leave this like this, it means that if in the future we want to return another error not listed here (like validate that this literal null can't be appended into the current state), we would be making a breaking change.

The way to solve it is by instead of listing all errors, we just list the AZ_OK and in the next like we add something like
`Non AZ_OK due to a failed while trying to append.

Reply:
Other APIs, like az_span already behave this way and I am following that pattern here. I would much rather discuss this offline and apply it across the board rather than discuss it per PR.

I don't fully agree with the conclusion here of not fully documenting the error states and providing relevant information to the caller.